### PR TITLE
Horizontal median for segmentation

### PR DESCRIPTION
**Pull request by knoellle**
## Introduced Changes

Instead of using a vertical moving median for segmenting vertical scan lines, the median is calculated by looking left and right of the considered pixels.
Since we have a horizontal stride of 4 anyways, we simply ignored most pixels.

This PR is part of a set:
- #683
- #684
- #686
- #687

 ~~Collect them all for a chance to win a golden PR! Buy now and win!~~
If you wish to test them all together, use this branch: [fliewatüüt](https://github.com/knoellle/hulk/tree/fliewat%C3%BC%C3%BCt)

Fixes #

## Ideas for Next Iterations (Not This PR)

Investigate also using horizontal median for calculating the color

## How to Test

Stare at segments, pray for salvation.

### CODE REVIEW

File: crates/vision/src/image_segmenter.rs [hunk 1]
```diff
@@ -225,19 +225,19 @@ fn new_vertical_scan_line(
...
         }
         MedianModeParameters::FivePixels => {
-            let second_previous_pixel = image.at(position, start_y - 2);
-            let previous_pixel = image.at(position, start_y - 1);
-            let next_pixel = image.at(position, start_y + 1);
-            let second_next_pixel = image.at(position, start_y + 2);
+            let second_previous_pixel = image.at(position - 2, start_y);
+            let previous_pixel = image.at(position - 1, start_y);
+            let next_pixel = image.at(position + 1, start_y);
+            let second_next_pixel = image.at(position + 2, start_y);
```
Comment (ichbinJosi) [inline crates/vision/src/image_segmenter.rs:240]: please update variables analogous

**Review by ichbinJosi — CHANGES_REQUESTED** [review]
:+1:

File: crates/vision/src/image_segmenter.rs [hunk 2]
```diff
@@ -225,19 +225,19 @@ fn new_vertical_scan_line(
     let luminance_value_of_first_pixel = match median_mode {
         MedianModeParameters::Disabled => first_pixel,
         MedianModeParameters::ThreePixels => {
-            let previous_pixel = image.at(position, start_y - 1);
-            let next_pixel = image.at(position, start_y + 1);
+            let previous_pixel = image.at(position - 1, start_y);
+            let next_pixel = image.at(position + 1, start_y);
```
[the remark this answers is not in the document]
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs]: I removed those variables entirely, using arrays now.

File: crates/vision/src/image_segmenter.rs (hunk 1 above)
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs]: Removed

File: crates/vision/src/image_segmenter.rs [hunk 3]
```diff
@@ -259,19 +259,19 @@ fn new_vertical_scan_line(
         let luminance_value = match median_mode {
             MedianModeParameters::Disabled => pixel,
             MedianModeParameters::ThreePixels => {
-                let previous_pixel = image.at(position, y - 1);
-                let next_pixel = image.at(position, y + 1);
+                let previous_pixel = image.at(position - 1, y);
+                let next_pixel = image.at(position + 1, y);
```
[the remark this answers is not in the document]
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs]: Removed

File: crates/vision/src/image_segmenter.rs [hunk 4]
```diff
@@ -259,19 +259,19 @@ fn new_vertical_scan_line(
...
             }
             MedianModeParameters::FivePixels => {
-                let second_previous_pixel = image.at(position, y - 2);
-                let previous_pixel = image.at(position, y - 1);
-                let next_pixel = image.at(position, y + 1);
-                let second_next_pixel = image.at(position, y + 2);
+                let second_previous_pixel = image.at(position - 2, y);
+                let previous_pixel = image.at(position - 1, y);
+                let next_pixel = image.at(position + 1, y);
+                let second_next_pixel = image.at(position + 2, y);
```
[the remark this answers is not in the document]
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs]: Removed

File: crates/vision/src/image_segmenter.rs [hunk 5]
```diff
@@ -190,8 +191,8 @@ fn median_of_three(first: u8, second: u8, third: u8) -> u8 {
     }
 }
 
-fn median_of_five(first: u8, second: u8, third: u8, fourth: u8, fifth: u8) -> u8 {
-    let mut values = [first, second, third, fourth, fifth];
+fn median_of_five(mut values: [u8; 5]) -> u8 {
+    // let mut values = [first, second, third, fourth, fifth];
```
Comment (ichbinJosi) [inline crates/vision/src/image_segmenter.rs:195]: Can you delete this comment?

File: crates/vision/src/image_segmenter.rs [hunk 6]
```diff
@@ -580,15 +563,15 @@ mod tests {
         let scan_line = new_vertical_scan_line(
             &image,
             &field_color,
-            0,
+            1,
```
Comment (ichbinJosi) [inline crates/vision/src/image_segmenter.rs:566]: Why did you change this value?

File: crates/vision/src/image_segmenter.rs [hunk 7]
```diff
@@ -743,27 +726,37 @@ mod tests {
...
+        let row = |y| {
+            [
+                YCbCr422::new(y, 0, 0, 0),
+                YCbCr422::new(y, 0, 0, 0),
+                YCbCr422::new(y, 0, 0, 0),
+            ]
+        };
         let image = YCbCr422Image::from_ycbcr_buffer(
-            1,
+            3,
```
Comment (ichbinJosi) [inline crates/vision/src/image_segmenter.rs:737]: And this one?

File: crates/vision/src/image_segmenter.rs [hunk 8]
```diff
@@ -886,28 +879,38 @@ mod tests {
...
             12,
-            vec![
+            [
                 // only evaluating every secondth 422 pixel
-                YCbCr422::new(3, 0, 0, 0),
-                YCbCr422::new(3, 0, 0, 0), // skipped
-                YCbCr422::new(2, 0, 0, 0),
+                row(3),
+                row(3), // skipped
+                row(2),
```
Comment (ichbinJosi) [inline crates/vision/src/image_segmenter.rs:896]: Why 9 lines with `YCbCr422::new(y, 0, 0, 0)` instead of three as previously?

**Review by ichbinJosi — CHANGES_REQUESTED** [review]

File: crates/vision/src/image_segmenter.rs (hunk 8 above)
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs:896]: The test image used to be 1x12 and now is 3x12 to accommodate taking a horizontal median of three.
I'm not sure where you're getting the 9 from.

File: crates/vision/src/image_segmenter.rs (hunk 6 above)
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs:566]: If we use the first column in the image, we can't take a horizontal median of three as the sample to the left would be at x = -1.

File: crates/vision/src/image_segmenter.rs (hunk 7 above)
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs:737]: Same thing here, you can't take a horizontal median of three on a 1 pixel wide image.

File: crates/vision/src/image_segmenter.rs (hunk 8 above)
Comment (schluis) [reply crates/vision/src/image_segmenter.rs:896]: Since row(y) now creates three values. And 3 * row = 3 * 3 = 9. But that makes sence since the image width is now three...

**Review by schluis — APPROVED** [review]
Nice 👌🏻

File: crates/vision/src/image_segmenter.rs (hunk 5 above)
Comment (knoellle) [reply crates/vision/src/image_segmenter.rs]: Removed

**Review by schluis — APPROVED** [review]
